src/EventStore.Core/DataStructures/ProbabilisticFilter/PersistentBloomFilters/BloomFilterAccessor.cs

```diff
@@ -0,0 +1,226 @@
+		public byte* Pointer {
+			get => _pointer;
+			set {
+				if ((long)value % CacheLineSize != 0) {
+					//qq consider whether to keep this.. throw? log error?
+					// do at least test out how the error message comes out
+					_log.Error("Pointer {pointer} was not aligned to a cacheline {cacheLineSize}", (long)value, CacheLineSize);
+					throw new InvalidOperationException($"Pointer {(long)value} was not aligned to a cacheline {CacheLineSize}");
+				}
+				_pointer = value;
+			}
+		}
+
+		// converts a bit position in the logical filter into a byte position in the file
+		public long GetBytePositionInFile(long bitPosition) {
+			var bytePositionInFile = bitPosition / 8;
+
+			// account for 4 bytes at the end of each cache line being a hash
+			bytePositionInFile += bytePositionInFile / (CacheLineSize - HashSize) * HashSize;
+
+			// first cache line is reserved for the header
+			bytePositionInFile += CacheLineSize;
+
+			return bytePositionInFile;
+		}
+
+		// converts page number into byte position of the page in the file
+		public (long PositionInFile, int PageSize) GetPagePositionInFile(int pageNumber) {
+			// first cache line is reserved for the header
+			var positionInFile = CacheLineSize + (pageNumber * PageSize);
+
+			var pageSize = Math.Min(PageSize, (int)(FileSize - positionInFile));
+
+			return (positionInFile, pageSize);
+		}
+
+		// determines which page a bytePosition in the file is for.
+		public long GetPageNumber(long bytePositionInFile) {
+			// pages start after the header
+			return (bytePositionInFile - CacheLineSize) / PageSize;
+		}
+
+		public void EnsureBounds(long bytePosition, int length) {
+			if (bytePosition > FileSize - length) {
+				throw new ArgumentOutOfRangeException(
+					nameof(bytePosition),
+					bytePosition,
+					$"Position {bytePosition:N0} + length {length:N0} = {bytePosition + length:N0} out of file size {FileSize:N0}");
+			}
+		}
+
+		public ref byte ReadByte(long bytePositionInFile) {
+			EnsureBounds(bytePositionInFile, length: 1);
+			return ref *(Pointer + bytePositionInFile);
+		}
+
+		public Span<byte> ReadCacheLineFor(long bytePositionInFile) {
+			bytePositionInFile = bytePositionInFile / CacheLineSize * CacheLineSize;
+			return ReadBytes(bytePositionInFile, CacheLineSize);
+		}
+
+		public Span<byte> ReadBytes(long bytePositionInFile, int count) {
+			EnsureBounds(bytePositionInFile, count);
+			return new Span<byte>(Pointer + bytePositionInFile, count);
+		}
+
+		public bool IsBitSet(long bitPosition) {
+			var bytePositionInFile = GetBytePositionInFile(bitPosition);
+			var byteValue = ReadByte(bytePositionInFile);
+			return byteValue.IsBitSet(bitPosition % 8);
+		}
+
+		public void SetBit(long bitPosition) {
+			var bytePositionInFile = GetBytePositionInFile(bitPosition);
+			ref var byteValue = ref ReadByte(bytePositionInFile);
+			byte oldByte = byteValue;
+			byteValue = byteValue.SetBit(bitPosition % 8);
+			var cacheLine = ReadCacheLineFor(bytePositionInFile);
```

Comment:
suggestion: also read the cache line and recompute the hash only if the byte's value has changed.

Reply:
ah yeah good point